File: dotnet/2024/AdventOfCode.Y2024/Day01.cs
using AdventOfCode.Lib;

namespace AdventOfCode.Y2024;

[AocPuzzle(2024, 1, "Historian Hysteria", "C#")]
public sealed class Day01 : IAocDay<int>
{
    public static int Part1(AocInput input) =>
        input.AllLines
        | (lines =>
        {
            var left = Instructions(lines, 0);
            var right = Instructions(lines, 1);
            return left.Zip(right, (l, r) => Math.Abs(l - r)).Sum();
        });

    public static int Part2(AocInput input) =>
        input.AllLines
        | (lines =>
        {
            var left = Instructions(lines, 0);
            var counts = Instructions(lines, 1).CountBy(i => i).ToDictionary();
            return left.Sum(id => counts.GetValueOrDefault(id) * id);
        });

    private static IEnumerable<int> Instructions(IEnumerable<string> lines, int col) =>
        from line in lines
        let nums = line.Split("   ").Select(int.Parse).ToArray()
        orderby nums[col]
        select nums[col];
}

File: python/y2024/day01.py
from aoc.core import solve, AocInput
from collections import Counter

def instructions(lines: list[str], col: int) -> list[int]:
    return sorted(int(line.split("   ")[col]) for line in lines)


def part1(input: AocInput) -> int:
    return sum(abs(l - r) for l, r in zip(instructions(input.lines, 0), instructions(input.lines, 1)))


def part2(input: AocInput) -> int:
    left = instructions(input.lines, 0)
    counts = Counter(instructions(input.lines, 1))
    return sum([counts[id] * id for id in left])


def main() -> None:
    YEAR = "y2024"
    DAY = "Day01"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1, 1660292)
    solve(YEAR, DAY, part2, 22776016)
